integration/helpers/deployer/utils/retry.py

```diff
@@ -38,3 +39,38 @@ def wrapper(*args, **kwargs):
         return wrapper
 
     return retry_wrapper
+
+
+def retry_with_exponential_backoff_and_jitter(exc, attempts=3, delay=0.05, exc_raise=Exception, exc_raise_msg=""):
+    """
+    Retry decorator which defaults to 3 attempts based on exponential backoff
+    and a delay of 50ms.
+    After retries are exhausted, a custom Exception and Error message are raised.
+
+    :param exc: Exception to be caught for retry
+    :param attempts: number of attempts before exception is allowed to be raised.
+    :param delay: an initial delay which will exponentially increase based on the retry attempt.
+    :param exc_raise: Final Exception to raise.
+    :param exc_raise_msg: Final message for the Exception to be raised.
+    :return:
+    """
+
+    def retry_wrapper(func):
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            remaining_attempts = attempts
+            retry_attempt = 1
+
+            while remaining_attempts >= 1:
+                try:
+                    return func(*args, **kwargs)
+                except exc:
+                    sleep_time = random.uniform(0, math.pow(2, retry_attempt) * delay)
```

Comment:
What's the reason for adding jitter?
Shouldn't sleep time be calculated by `delay +/- jitter` instead of a random int between 0 to max delay?
The current approach on a 10th retry can range from 0 to 512 seconds. This seems to be a large range and I'm not sure what that will accomplish.

Reply:
The Cloudformation's`describe_stack` api that we use to check the stack status has hard limit and cannot be changed. Since we run the tests in parallel(for example 10 thread in concurrent), we will have 10 calls in the same time(imagine a big call cluster in a small time frame). So we need to add jitter(add some randomness) to spread the calls over a large time frame to avoid the throttling. If we just use exponential back off, the call cluster will appear once again after exponential time and won't solve the issue, therefore we need the jitter.

Reply:
Why is the final delay generated with a random between 0 to max delay instead of plus minus a percentage of the max delay? 
In the example of 10th retry, it could have only 1 second delay which defeats the purpose of exponential backoff.

Reply:
Green as minimal, Red as maximum, and Purple as average.
This is the current implementation:
![chrome_f5Cv4MNo4W](https://user-images.githubusercontent.com/3464015/143146650-8584774c-faf1-4a57-9368-c05d66c4e8c6.png)

With a percentage based jitter:
![chrome_SX71SQbQRJ](https://user-images.githubusercontent.com/3464015/143146706-b1e8f40f-981a-438b-b82e-52be393025cd.png)

Reply:
Our goal is trying to spread the api call among the test time frame as even as possible(If we use delay + jitter we might have some time with no call). Thinking now how to do some simulation about this.

Reply:
Tracking:

Current implementation is following the blog post: https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/

One assumption the blog post makes is that the calls will continue to happen until the job is complete. However, in our case, we will fail the test if it fails after an x amount of retries. Although the total amount of "Work" is less, there could be a case where a job rolls low numbers for 5 times and fails the execution.
I think there is a better solution for our use case than either full jitter or equal jitter.

@mingkun2020 will work on a simulator for our use case specifically to test out different approaches.

This comment is not blocking.